File: algorithms/word_pattern.rb
# @param {String} pattern
# @param {String} str
# @return {Boolean}
def word_pattern(pattern, str)
  words = str.split
  return false if pattern.length != words.length
  hash = Hash.new()
  pattern.each_char.with_index do |c, i|
    hash[c] ||= words[i]
    return false if hash[c] != words[i]
    return false if (i > 0) and
      ((pattern[i] == pattern[i - 1]) ^ (words[i] == words[i - 1]))
  end
  return true
end

raise unless word_pattern('abba', 'a b b a')
raise if word_pattern("abba", "dog dog dog dog")
raise if word_pattern("abba", "dog dog dog")

File: algorithms/word_pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        words = str.split()
        if len(pattern) != len(words):
            return False
        hash_word_char = {}
        hash_char_word = {}
        for (char, word) in zip(pattern, words):
            hash_word_char.setdefault(word, char)
            if hash_word_char[word] != char:
                return False
            hash_char_word.setdefault(char, word)
            if hash_char_word[char] != word:
                return False
        return True

t = Solution()
assert t.wordPattern("abba", "dog cat cat dog")
assert not t.wordPattern("aaaa", "dog cat cat dog")
assert not t.wordPattern("abba", "dog dog dog dog")
assert not t.wordPattern("abba", "dog cat cat fish")
assert not t.wordPattern("aaa", "aa aa aa aa")
print("tests passed")
